libcst/_nodes/base.py

```diff
@@ -401,8 +401,8 @@ def deep_remove(
         )
 
     def with_deep_changes(
-        self: _CSTNodeSelfT, old_node: "CSTNode", **changes: Any
-    ) -> Union[_CSTNodeSelfT, "CSTNode"]:
+        self: _CSTNodeSelfT, old_node: CSTNodeT, **changes: Any
+    ) -> Union[_CSTNodeSelfT, CSTNodeT]:
```

Comment:
The type annotations here are confusing to me.
`_CSTNodeSelfT` and `CSTNodeT` seems refer to same thing, why do we need two names instead of one.
Also, while you're on typing here, can we also fix this inconsistency:
https://github.com/Instagram/LibCST/blob/a4124c91016a3430a6ce8bee0be0059989df8859/libcst/_nodes/base.py#L375-L378

Reply:
There is no typing inconsistency with `deep_replace`. We care about the type of `self`, and the type we are going to replace `old_node` with, because those are the two possible types that we might get back from the function. `old_node` just needs to be any `CSTNode`, but it doesn't factor into the return. Either we replace a child, and the return is the type of `self`, or `old_node` == `self`, so our return is whatever type `new_node` is, hence the union.

I think this current fix is actually broken and I unnecessarily widened the type. `with_changes` guarantees the same node as output, so even if `old_node` == `self`, the output is still `_CSTNodeSelfT`. I can make a fix to that here and change back `old_node` to be of type `CSTNode` as it was before, and fix the return to not be a union but instead just be `_CSTNodeSelfT`

Reply:
Yeah, return `_CSTNodeSelfT ` makes more sense to me.